### PR TITLE
BUG: support fused types in roll_min/max #12373

### LINKED ISSUE #12373: BUG: rolling functions raise ValueError on float32 data
pd version: v0.18.0rc1

minimal example:

```
import pandas as pd
import numpy as np
pd.DataFrame(np.arange(5, dtype=np.float32)).rolling(window=3).max()
```

stack trace:

```
<ipython-input-25-795ba6485799> in <module>()
----> 1 pd.DataFrame(np.arange(5, dtype=np.float32)).rolling(window=3).max()

/venv/lib/python3.4/site-packages/pandas/core/window.py in max(self, **kwargs)
    767     @Appender(_shared_docs['max'])
    768     def max(self, **kwargs):
--> 769         return super(Rolling, self).max(**kwargs)
    770 
    771     @Substitution(name='rolling')

/venv/lib/python3.4/site-packages/pandas/core/window.py in max(self, how, **kwargs)
    541         if self.freq is not None and how is None:
    542             how = 'max'
--> 543         return self._apply('roll_max', how=how, **kwargs)
    544 
    545     _shared_docs['min'] = dedent("""

/venv/lib/python3.4/site-packages/pandas/core/window.py in _apply(self, func, window, center, check_minp, how, **kwargs)
    474 
    475             if values.ndim > 1:
--> 476                 result = np.apply_along_axis(calc, self.axis, values)
    477             else:
    478                 result = calc(values)

/venv/lib/python3.4/site-packages/numpy/lib/shape_base.py in apply_along_axis(func1d, axis, arr, *args, **kwargs)
     89     outshape = asarray(arr.shape).take(indlist)
     90     i.put(indlist, ind)
---> 91     res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
     92     #  if res is a number, then we have a smaller output array
     93     if isscalar(res):

/venv/lib/python3.4/site-packages/pandas/core/window.py in calc(x)
    471 
    472                 def calc(x):
--> 473                     return func(x, window, min_periods=self.min_periods)
    474 
    475             if values.ndim > 1:

/venv/lib/python3.4/site-packages/pandas/core/window.py in func(arg, window, min_periods)
    458                 def func(arg, window, min_periods=None):
    459                     minp = check_minp(min_periods, window)
--> 460                     return cfunc(arg, window, minp, **kwargs)
    461 
    462             # calculation function

pandas/algos.pyx in pandas.algos.roll_max (pandas/algos.c:37363)()

ValueError: Buffer dtype mismatch, expected 'float64_t' but got 'float'

```

### PR DESCRIPTION
**Pull request by joshuastorck**
- Closes #12373
- Added test_rolling_min_max_types to test_window.py
- passes `git diff upstream/master | flake8 --diff`

Ran the performance tests with no changes detected:

$ asv continuous -b 'stat_op.*rolling' upstream/master GH12373
· Creating environments
· Discovering benchmarks
·· Uninstalling from py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
·· Building for py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..................................................
·· Installing into py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
· Running 18 total benchmarks (2 commits \* 1 environments \* 9 benchmarks)
[  0.00%] · For pandas commit hash 170fb27a:
[  0.00%] ·· Building for py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.......................................................
[  0.00%] ·· Benchmarking py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  5.56%] ··· Running stat_ops.stats_rolling_mean.time_rolling_kurt                                                                                                                                  11.39ms
[ 11.11%] ··· Running stat_ops.stats_rolling_mean.time_rolling_max                                                                                                                                    7.48ms
[ 16.67%] ··· Running stat_ops.stats_rolling_mean.time_rolling_mean                                                                                                                                   7.22ms
[ 22.22%] ··· Running stat_ops.stats_rolling_mean.time_rolling_median                                                                                                                               114.21ms
[ 27.78%] ··· Running stat_ops.stats_rolling_mean.time_rolling_min                                                                                                                                    7.55ms
[ 33.33%] ··· Running stat_ops.stats_rolling_mean.time_rolling_skew                                                                                                                                  11.35ms
[ 38.89%] ··· Running stat_ops.stats_rolling_mean.time_rolling_std                                                                                                                                    9.13ms
[ 44.44%] ··· Running stat_ops.stats_rolling_mean.time_rolling_sum                                                                                                                                    6.93ms
[ 50.00%] ··· Running stat_ops.stats_rolling_mean.time_rolling_var                                                                                                                                    7.98ms
[ 50.00%] · For pandas commit hash 56e285a6:
[ 50.00%] ·· Building for py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.......................................................
[ 50.00%] ·· Benchmarking py2.7-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 55.56%] ··· Running stat_ops.stats_rolling_mean.time_rolling_kurt                                                                                                                                  11.66ms
[ 61.11%] ··· Running stat_ops.stats_rolling_mean.time_rolling_max                                                                                                                                    8.68ms
[ 66.67%] ··· Running stat_ops.stats_rolling_mean.time_rolling_mean                                                                                                                                   7.48ms
[ 72.22%] ··· Running stat_ops.stats_rolling_mean.time_rolling_median                                                                                                                               107.71ms
[ 77.78%] ··· Running stat_ops.stats_rolling_mean.time_rolling_min                                                                                                                                    8.74ms
[ 83.33%] ··· Running stat_ops.stats_rolling_mean.time_rolling_skew                                                                                                                                  11.53ms
[ 88.89%] ··· Running stat_ops.stats_rolling_mean.time_rolling_std                                                                                                                                    9.25ms
[ 94.44%] ··· Running stat_ops.stats_rolling_mean.time_rolling_sum                                                                                                                                    6.97ms
[100.00%] ··· Running stat_ops.stats_rolling_mean.time_rolling_var                                                                                                                                    8.24msBENCHMARKS NOT SIGNIFICANTLY CHANGED.

### CODE REVIEW

File: pandas/algos.pyx [hunk 1]
```diff
@@ -1633,21 +1633,34 @@ from libc cimport stdlib
...
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def roll_min(ndarray[numeric] a, int window, int minp):
+    "Moving min of 1d array of any numeric type along axis=0 ignoring NaNs."
+    return _roll_min_max(a, window, minp, 0)
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _roll_min_max(ndarray[numeric] a, int window, int minp, bint is_max):
```
Comment (jreback) [inline pandas/algos.pyx]: this should be ``cdef``

**Comment by jreback** [comment]
ideally if you can run the perf tests that cover rolling min/max and see if anything is changing (prob not).

**Comment by jreback** [comment]
will be in a slight conflict with #12376, IOW we don't need to upcast for min/max (but do for other algos). So need a flag for that somewhere.

**Comment by joshuastorck** [comment]
I feel a bit embarrassed for not completely reading the issue and noticing that someone already did a PR. I think my fix is a bit more comprehensive. I’ll take a look this week and try to merge.

**Comment by jreback** [comment]
@joshuastorck no worries. I think the other fix has some comprehensive tests which is good and will move the needle. however your fix on top is ultimately better as the fused types are more general.

**Comment by BranYang** [comment]
@joshuastorck Your fix makes more sense than just force every roll_<function> to only work on float64. I was thinking about this but I am not very familiar with Cython so I just haven't touch the Cython code.
Glad to see that someone will do this for making pandas better

**Comment by jreback** [comment]
@joshuastorck can you rebase/update

**Comment by joshuastorck** [comment]
I'm closing this in favor of #12595

**Comment by jreback** [comment]
@joshuastorck ok!

**Pull request closed without merging**